kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1242 @@
+                    });
+                }
+                State::PaddingWrite => {
+                    // replace the buffer after the padding is written
+                    self.reset_process_loading_metadata(); // the final reset after the padding write callback
+                    self.buffer.replace(buffer);
+                }
+                State::Fail => {
+                    // If we failed at any of writing, we want to set the state to PaddingWrite
+                    // so that the callback after writing the padding app will get triggererd
+                    self.state.set(State::PaddingWrite);
+                    self.buffer.replace(buffer);
+                }
+                State::Setup => {
+                    // We have finished writing the post app padding
+                    self.buffer.replace(buffer);
+                    self.state.set(State::FirstWrite);
+                    // let the client know we are done setting up
+                    self.client.map(|client| {
+                        client.setup_done();
+                    });
+                }
+                State::Load => {
+                    // We finished writing pre-padding and we need to Load the app
+                    self.buffer.replace(buffer);
+                }
+            }
+        });
+    }
+}
+
+/// Interface exposed to the app_loader capsule
+impl<'a, C: 'static + Chip> DynamicProcessLoading for DynamicProcessLoader<'a, C> {
+    fn set_client(&self, client: &'static dyn DynamicProcessLoadingClient) {
+        self.client.set(client);
+    }
+
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode> {
+        //TODO(?): Check if it is a newer version of an existing app.
+        // We can potentially flash the new app and load it before erasing the old one.
+        // What happens to the process though? Need to delete it from the process array and load it back in?
+
+        let flash_start = self.flash.get().as_ptr() as usize; //start of the flash region
+        let flash = self.flash.get();
+
+        if self.state.is_none() {
+            self.state.set(State::Setup);
+            match self.find_next_available_address(flash, app_length) {
+                Ok(new_app_start_address) => {
+                    let offset = new_app_start_address - flash_start;
+                    let new_process_flash = self
+                        .flash
+                        .get()
+                        .get(offset..offset + app_length)
+                        .ok_or(ErrorCode::FAIL)?;
+
+                    self.new_process_flash.set(new_process_flash);
+                    self.new_app_start_addr.set(new_app_start_address);
+                    self.new_app_length.set(app_length);
+
+                    // check what kind of padding we have to write, no padding, pre padding, post padding or both pre and post padding
+                    self.check_padding_requirements();
```

Comment:
With check_padding_reqs integrated into find_next_available this code is a little more straightforward.

Reply:
so we compute the requirements in find_next_address, but still write the padding in setup right? I find it odd otherwise that the function trying to find the next address and padding requirement also writes the padding. It makes sense that the main setup logic should take care of this.